libp2p-identify/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+//! information B knows about A. Also includes the addresses B is listening on.
+//!
+//! When two nodes connect to each other, the listening half sends a message to the dialing half,
+//! indicating the information, and then the protocol stops.
+
+extern crate bytes;
+extern crate futures;
+extern crate multiaddr;
+extern crate libp2p_peerstore;
+extern crate libp2p_swarm;
+extern crate protobuf;
+extern crate tokio_io;
+
+use bytes::{Bytes, BytesMut};
+use futures::{Future, Stream, Sink};
+use libp2p_swarm::{ConnectionUpgrade, Endpoint};
+use multiaddr::Multiaddr;
+use protobuf::Message as ProtobufMessage;
+use protobuf::core::parse_from_bytes as protobuf_parse_from_bytes;
+use protobuf::repeated::RepeatedField;
+use std::io::{Error as IoError, ErrorKind as IoErrorKind};
+use std::iter;
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_io::codec::length_delimited;
+
+mod structs_proto;
+
+/// Prototype for an upgrade to the identity protocol.
+#[derive(Debug, Clone)]
+pub struct IdentifyProtocol {
+	pub information: IdentifyInfo,
+}
+
+/// Information sent from the listener to the dialer.
+#[derive(Debug, Clone)]
+pub struct IdentifyInfo {
+	/// Public key of the node.
+	pub public_key: Vec<u8>,
+	/// Version of the "global" protocol, eg. `ipfs/1.0.0`.
+	pub protocol_version: String,
+	/// Name and version. Can be thought as similar to the `User-Agent` header of HTTP.
+	pub agent_version: String,
+	/// Addresses that are listened on.
+	pub listen_addrs: Vec<Multiaddr>,
+	/// Address that the server uses to communicate with the dialer.
+	pub observed_addr: Multiaddr,
+	/// Protocols supported by the remote.
+	pub protocols: Vec<String>,
+}
+
+impl<C> ConnectionUpgrade<C> for IdentifyProtocol
+    where C: AsyncRead + AsyncWrite + 'static
+{
+	type NamesIter = iter::Once<(Bytes, Self::UpgradeIdentifier)>;
+	type UpgradeIdentifier = ();
+
+	#[inline]
+	fn protocol_names(&self) -> Self::NamesIter {
+		iter::once((Bytes::from("/ipfs/id/1.0.0"), ()))
+	}
+
+	type Output = Option<IdentifyInfo>;
+	type Future = Box<Future<Item = Self::Output, Error = IoError>>;
+
+	fn upgrade(self, socket: C, _: (), ty: Endpoint) -> Self::Future {
+		// TODO: use jack's varint library instead
+		let socket = length_delimited::Builder::new().length_field_length(1).new_framed(socket);
+
+		match ty {
+			Endpoint::Dialer => {
+				let future = socket.into_future()
+				                   .map(|(msg, _)| msg)
+				                   .map_err(|(err, _)| err)
+				                   .and_then(|msg| if let Some(msg) = msg {
+					Ok(Some(parse_proto_msg(msg)?))
+				} else {
+					Ok(None)
+				});
+
+				Box::new(future) as Box<_>
+			}
+
+			Endpoint::Listener => {
+				let info = self.information;
+
+				let listen_addrs = info.listen_addrs
+				                       .into_iter()
+				                       .map(|addr| addr.to_string().into_bytes())
+				                       .collect();
+
+				let mut message = structs_proto::Identify::new();
```

Comment:
Can we not use `serde` or suchlike? I know you don't like it but this is really ugly.

Reply:
I'm not against `serde`. The most mature Rust crate that supports protobuf is `protobuf`, and that's the API that they provide. We don't have much choice. I think it's also the only crate that lets you generate a definition at compile-time instead of parsing the messages structure at runtime.